File: [F]string/[F]string/8-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        sign = 0
        num = 0
        max_int = 2**31 - 1
        min_int = - (2**31)

        for c in s:
            if c == ' ' and not sign:
                continue
            elif sign == 0 and c == '+':
                sign = 1
            elif sign == 0 and c == '-':
                sign = - 1
            elif not c.isdigit():
                break
            else:
                if sign == 0:
                    sign = 1
                if sign * num > max_int // 10 or \
                   (sign * num == max_int // 10 and int(c) > max_int % 10):
                    return max_int
                if sign * num < int(min_int / 10) or \
                     (sign * num == int(min_int / 10) and int(c) > 10 - min_int % 10):
                    return min_int
                num = num * 10 + int(c)
            
        return sign * num
        
# time O(n), due to traverse
# space O(1)
# using string and handle value’s bound
'''
-25//10 == -3
int(-25/10) == -2
-102 % 10 == 8
2 == 10 - (-102 % 10)
'''

File: [F]string/[F]string/8-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        boolean metSign = false;
        boolean metNum = false;
        int sign = 1;
        int num = 0;
        int min = Integer.MIN_VALUE;
        int max = Integer.MAX_VALUE;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == ' ') {
                if (metSign || metNum) {
                    return sign * num;
                } else {
                    continue;
                }
            } else if (c == '-' || c == '+') {
                if (metSign || metNum) {
                    return sign * num;
                }
                metSign = true;
                if (c == '-') {
                    sign = - 1;
                }
            } else if (Character.isDigit(c)) {
                metNum = true;
                int val = c - '0';
                if (sign * num > max / 10 || (sign * num == max / 10 && sign * val > max % 10)) {
                    return max;
                } else if (sign * num < min / 10 || (sign * num == min / 10 && sign * val < min % 10)) {
                    return min;
                } else {
                    num = num * 10 + val;
                }
            } else {
                return sign * num;
            }
        }
        return sign * num;
    }
}

// time O(n), due to traverse
// space O(1)
// using string and handle value’s bound